hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/EightToSevenDowngradeHandler.java

```diff
@@ -161,19 +166,32 @@ static void unsetInitialVersion(HoodieTableConfig tableConfig, Map<ConfigPropert
     tableConfig.getProps().remove(HoodieTableConfig.INITIAL_VERSION.key());
   }
 
-  static void unsetRecordMergeMode(HoodieTableConfig tableConfig, Map<ConfigProperty, String> tablePropsToAdd) {
-    Triple<RecordMergeMode, String, String> mergingConfigs =
-        HoodieTableConfig.inferCorrectMergingBehavior(
-            tableConfig.getRecordMergeMode(), tableConfig.getPayloadClass(),
-            tableConfig.getRecordMergeStrategyId(), tableConfig.getPreCombineField(),
-            tableConfig.getTableVersion());
-    if (StringUtils.nonEmpty(mergingConfigs.getMiddle())) {
-      tablePropsToAdd.put(HoodieTableConfig.PAYLOAD_CLASS_NAME, mergingConfigs.getMiddle());
-    }
-    if (StringUtils.nonEmpty(mergingConfigs.getRight())) {
-      tablePropsToAdd.put(HoodieTableConfig.RECORD_MERGE_STRATEGY_ID, mergingConfigs.getRight());
+  static List<ConfigProperty> unsetRecordMergeMode(HoodieWriteConfig config, HoodieTableConfig tableConfig, Map<ConfigProperty, String> tablePropsToAdd) {
+    String payloadClass = tableConfig.getPayloadClass();
+    if (StringUtils.isNullOrEmpty(payloadClass)) {
+      RecordMergeMode mergeMode = tableConfig.getRecordMergeMode();
+      switch (mergeMode) {
+        case EVENT_TIME_ORDERING:
+          tablePropsToAdd.put(HoodieTableConfig.PAYLOAD_CLASS_NAME, DefaultHoodieRecordPayload.class.getName());
+          break;
+        case COMMIT_TIME_ORDERING:
+          tablePropsToAdd.put(HoodieTableConfig.PAYLOAD_CLASS_NAME, OverwriteWithLatestAvroPayload.class.getName());
+          break;
+        case CUSTOM:
+          //TODO: Handle CUSTOM payload
+          HoodieRecordMerger recordMerger = config.getRecordMerger();
+          if (recordMerger != null) {
+            throw new HoodieUpgradeDowngradeException(String.format("Custom payload class must be available when record merger is set %s", recordMerger.getClass().getName()));
+          }
```

Comment:
The custom merge mode downgrade logic is still pending. I think if there is custom merge mode available but no payload class, we should fail the downgrade since the payload class is not available to use for table version 6.

Reply:
I have changed the logic to throw exception for custom merge mode when payload class is not available. Should we use write config here to fetch the payload class?